### ISSUE TITLE
native function for partial trace (hacky implementation given)

### ISSUE THREAD

**Issue by rocketdude** [cvxpy/cvxpy#563]
Trying to use cvxpy for SDP in quantum information. It's great that the new cvxpy has built-in support for complex matrices, but we still need the partial trace function.

If ```rho``` is a matrix formed from tensor products:
<a href="https://www.codecogs.com/eqnedit.php?latex=\rho&space;=&space;X_1&space;\otimes&space;X_2&space;\otimes&space;\dots&space;\otimes&space;X_n" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\rho&space;=&space;X_1&space;\otimes&space;X_2&space;\otimes&space;\dots&space;\otimes&space;X_n" title="\rho = X_1 \otimes X_2 \otimes \dots \otimes X_n" /></a>
then the partial trace with respect to the k-th subsystem is:
<a href="https://www.codecogs.com/eqnedit.php?latex=\rho'&space;=&space;Tr(X_k)&space;\cdot&space;[&space;\right&space;]X_1&space;\otimes&space;X_2&space;\otimes&space;\dots&space;\otimes&space;X_{k-1}&space;\otimes&space;X_{k&plus;1}&space;\otimes&space;\dots&space;\otimes&space;X_n&space;]" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\rho'&space;=&space;Tr(X_k)&space;\cdot&space;[&space;\right&space;]X_1&space;\otimes&space;X_2&space;\otimes&space;\dots&space;\otimes&space;X_{k-1}&space;\otimes&space;X_{k&plus;1}&space;\otimes&space;\dots&space;\otimes&space;X_n&space;]" title="\rho' = Tr(X_k) \cdot [ \right ]X_1 \otimes X_2 \otimes \dots \otimes X_{k-1} \otimes X_{k+1} \otimes \dots \otimes X_n ]" /></a>
[Note: not all matrix ```rho``` can be written down as tensor products of smaller matrices.]

The simplest case would be if there are only two subsystems. For a matrix ```rho``` with only two subsystems, we can think of this matrix as ```N1 x N1``` block matrices ```M```. Each ```M``` has the size of ```N2 x N2```.
- Partial trace over the first subsystem is the sum of the block matrices ```M``` in the diagonal. This gives a matrix ```rho_primed``` of size ```N2 x N2```.
- Partial trace over the second subsystem performs trace over each block matrix ```M```. This gives a matrix ```rho_primed``` of size ```N1 x N1```.

How would you recommend one to perform partial trace on cvxpy? Any help is greatly appreciated. Thanks!

**Comment by rileyjmurray** [comment]
This question is more appropriate for stack exchange (with a cvxpy tag).

That said, here are some thoughts:
* At present, cvxpy Expressions can have at most two dimensions. So you would need to implement the tensor product as nested block 2d matrices in an appropriate basis.
* An expression of the form ``rho = X1 \tensor X2`` will be non-convex when ``X1`` and ``X2``  are variables. There is no way to work with expressions of that form with cvxpy, cvx, or any other platform using disciplined convex optimization.
* The definition you gave for the partial trace obscures the fact that it's a linear operator. In order to implement partial trace in cvxpy I think you will need the basis-dependent definition.

**Comment by rileyjmurray** [comment]
I think I got a little ahead of myself by closing this. Your particular application (with the tensor products and the quantum information) may be beyond the scope of  github issue, but if the partial trace is a useful linear operator, then it's reasonable to request that it be in cvxpy. I'll look into this.

**Comment by dbunandar** [comment]
Thanks a lot for looking into this. Yes, partial trace is a linear operator that is very useful in quantum information. With regards to your initial comments:

- Yes, we should represent this in two dimensions (as a matrix instead of a tensor). Since trace and partial trace are invariant in any basis, all we need to know is the dimension of each subsystem.
- We should start with ```rho``` being the variable, not ```X1``` nor ```X2```

It'd be awesome if partial trace is in cvxpy. Happy to discuss.

**Comment by rileyjmurray** [comment]
Okay if we start with ``rho`` and not ``X1,X2`` we can make this work.

Here's what I would recommend: figure out how to do this with numpy. If you can perform a partial trace using the helper functions in numpy, then you can perform it pretty easily in cvxpy. In particular if you write the code in numpy, then I would happy to reformat it to work well with cvxpy (i.e. reformat it to prevent performance / speed issues based on peculiarities of cvxpy's backend).

From there we'll see what we can do about making it a default cvxpy function.

Does that work?

**Comment by dbunandar** [comment]
Yes, that works. Here's how I would do it in numpy. Let me know if any part of the code doesn't make sense.

```python
import numpy as np


def partial_trace(rho, dims, axis=0):
    """
    Takes partial trace over the subsystem defined by 'axis'
    rho: a matrix
    dims: a list containing the dimension of each subsystem
    axis: the index of the subsytem to be traced out
    (We assume that each subsystem is square)
    """
    dims_ = np.array(dims)
    # Reshape the matrix into a tensor with the following shape:
    # [dim_0, dim_1, ..., dim_n, dim_0, dim_1, ..., dim_n]
    # Each subsystem gets one index for its row and another one for its column
    reshaped_rho = rho.reshape(np.concatenate((dims_, dims_), axis=None))

    # Move the subsystems to be traced towards the end
    reshaped_rho = np.moveaxis(reshaped_rho, axis, -1)
    reshaped_rho = np.moveaxis(reshaped_rho, len(dims)+axis-1, -1)

    # Trace over the very last row and column indices
    traced_out_rho = np.trace(reshaped_rho, axis1=-2, axis2=-1)

    # traced_out_rho is still in the shape of a tensor
    # Reshape back to a matrix
    dims_untraced = np.delete(dims_, axis)
    rho_dim = np.prod(dims_untraced)
    return traced_out_rho.reshape([rho_dim, rho_dim])


"""
Test out the partial_trace numpy module by creating a matrix
rho_ABC = rho_A \otimes rho_B \otimes rho_C
Each rho_i is normalized, i.e. Tr(rho_i) = 1
"""

# Generate the results we want
rho_A = np.random.rand(4, 4)
rho_A /= np.trace(rho_A)
rho_B = np.random.rand(3, 3)
rho_B /= np.trace(rho_B)
rho_C = np.random.rand(2, 2)
rho_C /= np.trace(rho_C)

rho_AB = np.kron(rho_A, rho_B)
rho_BC = np.kron(rho_B, rho_C)
rho_AC = np.kron(rho_A, rho_C)
rho_ABC = np.kron(rho_AB, rho_C)

# Try to get the results by doing partial_trace's from rho_ABC
rho_AB_test = partial_trace(rho_ABC, [4, 3, 2], axis=2)
rho_AC_test = partial_trace(rho_ABC, [4, 3, 2], axis=1)
rho_A_test = partial_trace(rho_AB, [4, 3], axis=1)
rho_B_test = partial_trace(rho_AB, [4, 3], axis=0)
rho_C_test = partial_trace(rho_AC, [4, 2], axis=0)

# See if the outputs of partial_trace are correct
print("rho_AB test correct? ", np.allclose(rho_AB_test, rho_AB))
print("rho_AC test correct? ", np.allclose(rho_AC_test, rho_AC))
print("rho_A test correct? ", np.allclose(rho_A_test, rho_A))
print("rho_B test correct? ", np.allclose(rho_B_test, rho_B))
print("rho_C test correct? ", np.allclose(rho_C_test, rho_C))
```

**Comment by rileyjmurray** [comment]
Okay that code is wonderfully general. Although, its use of multidimensional numpy arrays does complicate things. Here's *a* workaround (not a permanent one I think). The idea in the code below is to create a numpy array with identical entries as a cvxpy Variable, to perform operations in numpy's backend using the numpy representation, and then to reduce the final 2d numpy array to a cvxpy Expression. It's similar to what I did in #457.

The main drawback with this solution is that your resulting cvxpy Problems may take a long time to compile when you're working with moderate sized matrix variables. Can you try it out and let me know if you run into slow compile-time issues?

```
import numpy as np
import cvxpy
from cvxpy.expressions.expression import Expression


def expr_as_np_array(cvx_expr):
    if cvx_expr.is_scalar():
        return np.array(cvx_expr)
    elif len(cvx_expr.shape) == 1:
        return np.array([v for v in cvx_expr])
    else:
        # then cvx_expr is a 2d array
        rows = []
        for i in range(cvx_expr.shape[0]):
            row = [cvx_expr[i,j] for j in range(cvx_expr.shape[1])]
            rows.append(row)
        arr = np.array(rows)
        return arr


def np_array_as_expr(np_arr):
    aslist = np_arr.tolist()
    expr = cvxpy.bmat(aslist)
    return expr


def np_partial_trace(rho, dims, axis=0):
    """
    Takes partial trace over the subsystem defined by 'axis'
    rho: a matrix
    dims: a list containing the dimension of each subsystem
    axis: the index of the subsytem to be traced out
    (We assume that each subsystem is square)
    """
    dims_ = np.array(dims)
    # Reshape the matrix into a tensor with the following shape:
    # [dim_0, dim_1, ..., dim_n, dim_0, dim_1, ..., dim_n]
    # Each subsystem gets one index for its row and another one for its column
    reshaped_rho = np.reshape(rho, np.concatenate((dims_, dims_), axis=None))

    # Move the subsystems to be traced towards the end
    reshaped_rho = np.moveaxis(reshaped_rho, axis, -1)
    reshaped_rho = np.moveaxis(reshaped_rho, len(dims)+axis-1, -1)

    # Trace over the very last row and column indices
    traced_out_rho = np.trace(reshaped_rho, axis1=-2, axis2=-1)

    # traced_out_rho is still in the shape of a tensor
    # Reshape back to a matrix
    dims_untraced = np.delete(dims_, axis)
    rho_dim = np.prod(dims_untraced)
    return traced_out_rho.reshape([rho_dim, rho_dim])


def partial_trace(rho, dims, axis=0):
    if not isinstance(rho, Expression):
        rho = cvxpy.Constant(shape=rho.shape, value=rho)
    rho_np = expr_as_np_array(rho)
    traced_rho = np_partial_trace(rho_np, dims, axis)
    traced_rho = np_array_as_expr(traced_rho)
    return traced_rho



"""
Test out the partial_trace numpy module by creating a matrix
rho_ABC = rho_A \otimes rho_B \otimes rho_C
Each rho_i is normalized, i.e. Tr(rho_i) = 1
"""

# Generate five test cases
rho_A = np.random.rand(4, 4) + 1j*np.random.rand(4, 4)
rho_A /= np.trace(rho_A)
rho_B = np.random.rand(3, 3) + 1j*np.random.rand(3, 3)
rho_B /= np.trace(rho_B)
rho_C = np.random.rand(2, 2) + 1j*np.random.rand(2, 2)
rho_C /= np.trace(rho_C)
rho_AB = np.kron(rho_A, rho_B)
rho_AC = np.kron(rho_A, rho_C)

# Construct a cvxpy Variable with value equal to rho_A \otimes rho_B \otimes rho_C.
temp = np.kron(rho_AB, rho_C)
rho_ABC = cvxpy.Variable(shape=temp.shape, complex=True)
rho_ABC.value = temp

# Try to recover simpler tensors products by taking partial traces of
# more complicated tensors.
rho_AB_test = partial_trace(rho_ABC, [4, 3, 2], axis=2)
rho_AC_test = partial_trace(rho_ABC, [4, 3, 2], axis=1)
rho_A_test = partial_trace(rho_AB_test, [4, 3], axis=1)
rho_B_test = partial_trace(rho_AB_test, [4, 3], axis=0)
rho_C_test = partial_trace(rho_AC_test, [4, 2], axis=0)

# See if the outputs of partial_trace are correct
print("rho_AB test correct? ", np.allclose(rho_AB_test.value, rho_AB))
print("rho_AC test correct? ", np.allclose(rho_AC_test.value, rho_AC))
print("rho_A test correct? ", np.allclose(rho_A_test.value, rho_A))
print("rho_B test correct? ", np.allclose(rho_B_test.value, rho_B))
print("rho_C test correct? ", np.allclose(rho_C_test.value, rho_C))
```

**Comment by rileyjmurray** [comment]
Update:

I've tried a different (read, less-hacky) approach than the one I gave above, but it was too slow. I think that the implementation given in my previous comment should be the go-to for partial trace, at least for now.

Before we make partial trace part of cvxpy we'll need to have native support for Expressions with ``.shape`` of length greater than 2.

**Comment by dbunandar** [comment]
Got it. So far it's been working for my small problems. Thank you for your help!

**Comment by SteveDiamond** [comment]
This implementation might be good to use: https://github.com/jump-dev/Convex.jl/blob/4068f80ec581ae77d8000d7f3fcb0387ce306577/src/atoms/affine/partialtrace.jl

**Comment by SteveDiamond** [comment]
I converted the Julia to Python. It's definitely faster than the naive implementation above. I will do a PR with this later.
```
def _term(rho, j, dims, axis=0):
    # (I ⊗ <j| ⊗ I) x (I ⊗ |j> ⊗ I) for all j's
    # in the system we want to trace out.
    # This function returns the jth term in the sum, namely
    # (I ⊗ <j| ⊗ I) x (I ⊗ |j> ⊗ I).
    a = sp.coo_matrix(([1.0], ([0], [0])))
    b = sp.coo_matrix(([1.0], ([0], [0])))
    for (i_axis, dim) in enumerate(dims):
        if i_axis == axis:
            v = sp.coo_matrix(([1], ([j], [0])), shape=(dim, 1))
            a = sp.kron(a, v.T)
            b = sp.kron(b, v)
        else:
            eye_mat = sp.eye(dim)
            a = sp.kron(a, eye_mat)
            b = sp.kron(b, eye_mat)
    return a @ rho @ b

def partial_trace(rho, dims, axis=0):
    rho = Expression.cast_to_const(rho)
    if rho.ndim < 2 or rho.shape[0] != rho.shape[1]:
        raise ValueError("Only supports square matrices.")
    if axis < 0 or axis >= len(dims):
        raise ValueError(f"Invalid axis argument, should be between 0 and {len(dims)}, got {axis}")
    if rho.shape[0] != np.prod(dims):
        raise ValueError("Dimension of system doesn't correspond to dimension of subsystems")
    return sum([_term(rho, j, dims, axis) for j in range(dims[axis])])
```

**Comment by duguyipiao** [comment]
> I converted the Julia to Python. It's definitely faster than the naive implementation above. I will do a PR with this later.
> 
> ```
> def _term(rho, j, dims, axis=0):
>     # (I ⊗ <j| ⊗ I) x (I ⊗ |j> ⊗ I) for all j's
>     # in the system we want to trace out.
>     # This function returns the jth term in the sum, namely
>     # (I ⊗ <j| ⊗ I) x (I ⊗ |j> ⊗ I).
>     a = sp.coo_matrix(([1.0], ([0], [0])))
>     b = sp.coo_matrix(([1.0], ([0], [0])))
>     for (i_axis, dim) in enumerate(dims):
>         if i_axis == axis:
>             v = sp.coo_matrix(([1], ([j], [0])), shape=(dim, 1))
>             a = sp.kron(a, v.T)
>             b = sp.kron(b, v)
>         else:
>             eye_mat = sp.eye(dim)
>             a = sp.kron(a, eye_mat)
>             b = sp.kron(b, eye_mat)
>     return a @ rho @ b
> 
> def partial_trace(rho, dims, axis=0):
>     rho = Expression.cast_to_const(rho)
>     if rho.ndim < 2 or rho.shape[0] != rho.shape[1]:
>         raise ValueError("Only supports square matrices.")
>     if axis < 0 or axis >= len(dims):
>         raise ValueError(f"Invalid axis argument, should be between 0 and {len(dims)}, got {axis}.")
>     if rho.shape[0] != np.prod(dims):
>         raise ValueError("Dimension of system doesn't correspond to dimension of subsystems.")
>     return sum([_term(rho, j, dims, axis) for j in range(dims[axis])])
> ```
> 
> Edit: In the evaluation code above with dimensions 8, 6, 4, the Convex.jl partial_trace takes 0.6 seconds while the naive version takes 60 seconds.

Can you also take a look at the related function for "partial transpose"?
https://github.com/jump-dev/Convex.jl/blob/4068f80ec581ae77d8000d7f3fcb0387ce306577/src/atoms/affine/partialtranspose.jl

**Comment by duguyipiao** [comment]
If you prefer, partial_transpose can be implemented in a similar way:

```python
import numpy as np
import scipy.sparse as sp
import cvxpy
from cvxpy.expressions.expression import Expression

def _term(rho, i, j, dims, axis=0):
    # (I ⊗ |i><j| ⊗ I) x (I ⊗ |i><j| ⊗ I) for all (i,j)'s
    # in the system we want to transpose.
    # This function returns the (i,j)-th term in the sum, namely
    # (I ⊗ |i><j| ⊗ I) x (I ⊗ |i><j| ⊗ I).
    a = sp.coo_matrix(([1.0], ([0], [0])))
    for (i_axis, dim) in enumerate(dims):
        if i_axis == axis:
            v = sp.coo_matrix(([1], ([i], [j])), shape=(dim, dim))
            a = sp.kron(a, v)
        else:
            eye_mat = sp.eye(dim)
            a = sp.kron(a, eye_mat)
    return a @ rho @ a

def partial_transpose(rho, dims, axis=0):
    rho = Expression.cast_to_const(rho)
    if rho.ndim < 2 or rho.shape[0] != rho.shape[1]:
        raise ValueError("Only supports square matrices.")
    if axis < 0 or axis >= len(dims):
        raise ValueError(f"Invalid axis argument, should be between 0 and {len(dims)}, got {axis}.")
    if rho.shape[0] != np.prod(dims):
        raise ValueError("Dimension of system doesn't correspond to dimension of subsystems.")
    return sum([_term(rho, i, j, dims, axis) for i in range(dims[axis]) for j in range(dims[axis])])
```

For test:
```python
"""
Test out the partial_transpose numpy module by creating a matrix
rho_ABC = rho_A \otimes rho_B \otimes rho_C
Each rho_i is normalized, i.e. Tr(rho_i) = 1
"""

# Generate three test cases
rho_A = np.random.rand(8, 8) + 1j*np.random.rand(8, 8)
rho_A /= np.trace(rho_A)
rho_B = np.random.rand(6, 6) + 1j*np.random.rand(6, 6)
rho_B /= np.trace(rho_B)
rho_C = np.random.rand(4, 4) + 1j*np.random.rand(4, 4)
rho_C /= np.trace(rho_C)

rho_TC = np.kron(np.kron(rho_A, rho_B), rho_C.T)
rho_TB = np.kron(np.kron(rho_A, rho_B.T), rho_C)
rho_TA = np.kron(np.kron(rho_A.T, rho_B), rho_C)

# Construct a cvxpy Variable with value equal to rho_A \otimes rho_B \otimes rho_C.
temp = np.kron(np.kron(rho_A, rho_B), rho_C)
rho_ABC = cvxpy.Variable(shape=temp.shape, complex=True)
rho_ABC.value = temp

# Try to recover simpler tensors products by taking partial transposes of
# more complicated tensors.
rho_TC_test = partial_transpose(rho_ABC, [8, 6, 4], axis=2)
rho_TB_test = partial_transpose(rho_ABC, [8, 6, 4], axis=1)
rho_TA_test = partial_transpose(rho_ABC, [8, 6, 4], axis=0)

# See if the outputs of partial_transpose are correct
print("rho_TC test correct? ", np.allclose(rho_TC_test.value, rho_TC))
print("rho_TB test correct? ", np.allclose(rho_TB_test.value, rho_TB))
print("rho_TA test correct? ", np.allclose(rho_TA_test.value, rho_TA))

```

**Comment by SteveDiamond** [comment]
Thanks @duguyipiao! I'll add this into my PR (#1606).

**Comment by SteveDiamond** [comment]
Actually @duguyipiao I have a question. Is partial transpose supposed to be a hermitian transpose, or transpose as above?

**Comment by duguyipiao** [comment]
> Actually @duguyipiao I have a question. Is partial transpose supposed to be a hermitian transpose, or transpose as above?

Hi @SteveDiamond , it is NOT hermitian transpose, just transpose as above. See for example:
https://en.wikipedia.org/wiki/Peres%E2%80%93Horodecki_criterion

**Comment by rileyjmurray** [comment]
Resolved by #1606
